design/k8s-job-hook.md

```diff
@@ -0,0 +1,176 @@
+    properties:
+      image: "registry/lakefs-hooks:4"
+      command: ["python"]
+      args: ["bump-version.py"]
+      env:
+      - name: REPOSITORY
+        value: customers
+      - name: PROJECT
+        value: alpha
+```
+
+In this example we specified a post merge hook to execute a job. The job will use the image `next/lakefs-hooks:4` with the command python with the argument `bump-version.py`.
+The container will have the environment variables populated with REPOSITORY and PROJECT as defined by the user.
+
+The following environment variables will be also populated with the event information:
+
+```
+LAKEFS_HOOK_EVENTTYPE - Type of the event that triggered the action
+LAKEFS_HOOK_EVENTTIME - Time of the event that triggered
+LAKEFS_HOOK_ACTIONNAME - Containing Hook Action's name
+LAKEFS_HOOK_HOOKID - ID of the hook
+LAKEFS_HOOK_REPOSITORYID - ID of the repository
+LAKEFS_HOOK_BRANCHID - ID of the branch
+LAKEFS_HOOK_SOURCEREF - Reference to the source that triggered the event
+LAKEFS_HOOK_COMMITMESSAGE - The message for the commit
+LAKEFS_HOOK_COMMITTER - Name of the committer
+LAKEFS_HOOK_COMMIT_METADATA - Commit metadata (json serialized string)
+```
+
+By default lakeFS will use the following as base definition to schedule a job:
+
+```yaml
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: ""
+spec:
+  template:
+    spec:
+      restartPolicy: Never
+      containers:
+      - name: hook
+        image: ""
+        command: []
+        args: []
+```
+
+Note that the metadata.name, image, command and args will be set by lakeFS.
+Name - unique identifier to identify the specific job execution
+Image, command and args - set based on the hook information
+In case we would like to set a different job specification, we can specify in the lakeFS configuration (lakefs.yaml):
+
+```yaml
+hooks:
+  k8s-job:
+    spec_files:
+      default: /etc/lakefs/job_spec.yaml
+```
+
+Following an example of a `job_spec.yaml`:
+
+```yaml
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: ""
+  namespace: lakefs-hooks
+  labels:
+    app.kubernetes.io/name: post-merge-hook
+    app.kubernetes.io/version: "1.0.0"
+spec:
+  template:
+    metadata:
+      labels:
+        app.kubernetes.io/name: post-merge-hook
+        app.kubernetes.io/version: "1.0.0"
+    spec:
+      restartPolicy: Never
+      containers:
+        - name: hook
+          image: ""
+          command: []
+          args: []
+          env:
+            - name: SPECIAL_ENV
+              value: special_value
+          resources:
+            limits:
+              cpu: "2"
+              memory: 4G
+            requests:
+              cpu: "1"
+              memory: 2G
+```
+
+We can mount the job spec using a configmap having lakeFS use it as the base to the job template used for all k8s-job hooks.
+
+*Limit the end-user image use*
+
+Using the lakeFS configuration, we can specify a list of allowed images that the end-user and use in the k8s-job hook:
```

Comment:
Again, not sure this is needed. But if you decide to keep this, let's start by explaining the problem: You don't want to allow lakeFS to run any image, therefore you can limit it with this configuration.

Reply:
We should enable a way to run different types of images.
If we do not enable to specify it in the action - we will need to support multiple specs.
Because  spec files are part of the lakeFS configuration, it means we will need to redeploy.

Reply:
This is not a good way of handling user permissions.  E.g. it is very easy to break out of many images.

But I don't think we need to do this!  lakeFS can just specify what K8s role to use for the job.  The K8s admin can specify what that role is allowed to do.